draft-ietf-quic-transport.md

```diff
@@ -924,11 +924,19 @@ selected by the client, both to ensure correct routing toward the client and to
 allow the client to validate that the packet is in response to an Initial
 packet.
 
-A zero-length connection ID MAY be used when the connection ID is not needed for
-routing and the address/port tuple of packets is sufficient to identify a
-connection. An endpoint whose peer has selected a zero-length connection ID MUST
-continue to use a zero-length connection ID for the lifetime of the connection
-and MUST NOT send packets from any other local address.
+A zero-length connection ID can be used when a connection ID is not needed
+to route to the correct endpoint. An endpoint SHOULD NOT use a zero-length
+connection ID unless it can use only its IP address and port to identify a
+connection. The IP address and port used by a peer cannot be used for routing
+or connection identification as these values can change during a connection's
+lifetime, and the peer can reuse a given address and port for additional
+connections. Similarly, the peer's connection IDs cannot be used for routing
+or identification, as they are not transmitted in the short header packets
+they send. Note that multiplexing while using zero-length connection IDs and
+relying on the four-tuple of IP addresses and ports for routing will cause
+failures in the presence of connection migration, NAT rebinding, and client
+port reuse; and therefore MUST NOT be done unless an endpoint is certain that
+those protocol features are not in use.
```

Comment:
I think it might be a good idea to move the "MUST NOT unless" up front, then explain the cases where a zero-length connection ID can be used. E.g.,

"Multiplexing while using zero-length connection IDs and relying on the four-tuple of IP addresses and ports for routing will cause failures in the presence of connection migration, NAT rebinding, and client port reuse; and therefore MUST NOT be done unless an endpoint is certain that those protocol features are not in use. A zero-length connection ID can be used when a connection ID is not needed to route to the correct endpoint. An endpoint SHOULD NOT use a zero-length connection ID unless it can use only its IP address and port to identify a connection. ..."

Reply:
I think that becomes harder to read, but I think that's editorial so I'll let the editors decide how best to write this if we have consensus on the concept

Reply:
Yes, editorial, but what about just this -- it captures to normative meaning without being redundant with other parts of the spec.

_A zero-length connection ID can be used when a connection ID is not needed to route to the correct endpoint. Since relying on the IP address and port of the peer for routing or connection identification will cause failures in the presence of connection migration, NAT rebinding, and client port reuse for multiple connections, an endpoint MUST NOT do so unless it is certain that those protocol features are not used by the peer._

Reply:
I agree with @igorlord that there's redundancy here. Specifically, "endpoint SHOULD NOT use a zero-length connection ID unless it can use only its IP address and port" is at odds with "multiplexing while using zero-length connection IDs [...] MUST NOT be done". I've suggested fixes above.

Reply:
Removed the redundancy